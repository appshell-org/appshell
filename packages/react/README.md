<div align="center">
  <a href="https://github.com/navaris/appshell">
    <picture>
      <source media="(prefers-color-scheme: dark)" srcset="https://github.com/navaris/appshell/blob/main/assets/branding/appshell-logo-white_2x.png">
      <img alt="appshell" src="https://github.com/navaris/appshell/blob/main/assets/branding/appshell-logo_2x.png">
    </picture>
  </a>
</div>

[![Appshell CI](https://github.com/navaris/appshell/actions/workflows/pipeline.yml/badge.svg)](https://github.com/navaris/appshell/actions/workflows/pipeline.yml)

# @appshell/react

React utilites for building micro-frontends with Webpack Module federation and Appshell.

Working examples can be found [here](https://github.com/navaris/appshell/tree/main/examples).

## Getting Started

To begin, you'll need to install `@appshell/react`:

```console
npm install @appshell/react --save-dev
```

or

```console
yarn add -D @appshell/react
```

or

```console
pnpm add -D @appshell/react
```

## FederatedComponent

React component that dynamically loads federated components.

```tsx
import { FederatedComponent, ManifestProvider } from '@appshell/react';

<App>
  <ManifestProvider manifest={manifest}>
    <FederatedComponent remote="PingModule/Ping">
    <FederatedComponent remote="PongModule/Pong">
  </ManifestProvider>
</App>
```

## useRegistry

For access to the registry index.

```tsx
import { RegistryProvider, useRegistry } from '@appshell/react';

const MyComponent = () => {
  const index = useRegistry();

  ...
}

<RegistryProvider index={index}>
  <MyComponent />
</RegistryProvider>
```

## useMetadata

For access to the registry metadata.

```tsx
import { MetadataProvider, useMetadata } from '@appshell/react';

const MyComponent = () => {
  const metadata = useMetadata();

  ...
}

<MetadataProvider metadata={metadata}>
  <MyComponent />
</MetadataProvider>
```

## useManifest

For access to the manifest.

```tsx
import { ManifestProvider, useManifest } from '@appshell/react';

const MyComponent = () => {
  const manifest = useManifest();

  ...
}

<ManifestProvider manifest={manifest}>
  <MyComponent />
</ManifestProvider>
```

**Where does the registry come from?**

Currently, the registry is a shared directory or mounted volume. It does support calling an http endpoint, but it has not yet been implemented.

> See [@appshell/cli](https://www.npmjs.com/package/@appshell/cli)

Each micro-frontend registers its manifest with the registry

```bash
appshell register --manifest dist/appshell.manifest.json
```
