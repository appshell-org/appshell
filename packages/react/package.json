{
  "name": "@appshell/react",
  "version": "0.3.1-alpha.0",
  "description": "React utilities for building micro-frontends with Appshell and Module Federation",
  "main": "dist/main.js",
  "types": "dist/types/react/src/index.d.ts",
  "repository": "https://github.com/navaris/appshell.git",
  "author": "Robert Hamilton <rh@navaris.com>",
  "publishConfig": {
    "access": "public",
    "registry": "https://registry.npmjs.org/"
  },
  "scripts": {
    "build": "webpack --mode production && npm run build:types",
    "build:types": "tsc --emitDeclarationOnly",
    "clean": "rm -rf dist",
    "start": "webpack --watch --mode development --stats-error-details",
    "stats": "webpack --json > stats.json"
  },
  "keywords": [
    "module federation",
    "micro-frontends",
    "appshell"
  ],
  "license": "MIT",
  "peerDependencies": {
    "react": "^18.2.0",
    "react-dom": "^18.2.0"
  }
}
