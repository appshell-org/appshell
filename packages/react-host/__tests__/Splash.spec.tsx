/** @jest-environment jsdom */
/* eslint-disable no-underscore-dangle */
import { render } from '@testing-library/react';
import React from 'react';

declare global {
  interface Window {
    __appshell_env__: Record<string, string | undefined>;
  }
}

Object.defineProperty(window, '__appshell_env__', {
  value: {
    APPSHELL_ROOT: 'TestModule/TestComponent',
    APPSHELL_THEME_COLOR: '#000000',
    APPSHELL_PRIMARY_COLOR: 'orangered',
  },
});
describe('Splash', () => {
  let Splash;

  beforeEach(() => {
    // eslint-disable-next-line global-require, @typescript-eslint/no-var-requires
    Splash = require('../src/components/Splash').default;
  });

  it('should match snapshot', () => {
    const view = render(<Splash />);

    expect(view).toMatchSnapshot();
  });
});
