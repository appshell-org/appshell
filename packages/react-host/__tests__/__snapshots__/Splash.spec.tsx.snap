// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Splash should match snapshot 1`] = `
<div>
  <div
    style="display: flex; align-items: center; justify-content: center; flex-grow: 1; height: 100vh; background-color: rgb(0, 0, 0);"
  >
    <span
      style="display: inherit; position: relative;"
    >
      <span
        style="position: absolute; font-size: 8px; width: 24px; height: 24px; background: orangered; border-radius: 50%; animation: react-spinners-PropagateLoader-propogate-0 1s infinite; animation-fill-mode: forwards;"
      />
      <span
        style="position: absolute; font-size: 8px; width: 24px; height: 24px; background: orangered; border-radius: 50%; animation: react-spinners-PropagateLoader-propogate-1 1s infinite; animation-fill-mode: forwards;"
      />
      <span
        style="position: absolute; font-size: 8px; width: 24px; height: 24px; background: orangered; border-radius: 50%; animation: react-spinners-PropagateLoader-propogate-2 1s infinite; animation-fill-mode: forwards;"
      />
      <span
        style="position: absolute; font-size: 8px; width: 24px; height: 24px; background: orangered; border-radius: 50%; animation: react-spinners-PropagateLoader-propogate-3 1s infinite; animation-fill-mode: forwards;"
      />
      <span
        style="position: absolute; font-size: 8px; width: 24px; height: 24px; background: orangered; border-radius: 50%; animation: react-spinners-PropagateLoader-propogate-4 1s infinite; animation-fill-mode: forwards;"
      />
      <span
        style="position: absolute; font-size: 8px; width: 24px; height: 24px; background: orangered; border-radius: 50%; animation: react-spinners-PropagateLoader-propogate-5 1s infinite; animation-fill-mode: forwards;"
      />
    </span>
  </div>
</div>
`;
